File: app/src/main/java/system/StorageService.java
package system;




import domain.*;
import validators.*;

import java.time.Instant;
import java.util.*;
import java.util.concurrent.atomic.AtomicLong;
import java.util.stream.Collectors;

public class StorageService {

    private final Set<container> containers = new HashSet<>();
    private final Set<slot> slots = new HashSet<>();
    private final Set<placement> placements = new HashSet<>();
    private final Set<sample> sample = new HashSet<>();

    // Генератор  ID
    private final AtomicLong idGenerator = new AtomicLong(1);

    public long addContainer(String name, ContainerType type, String owner) {
        ContainerValidator.validateName(name);
        ContainerValidator.validateType(type);
        container c = new container(idGenerator.getAndIncrement(), name.trim(), type, owner);
        containers.add(c);
        return c.getId();
    }

    public container getContainer(long id) {
        return containers.stream().filter(c -> c.getId() == id).findFirst().orElse(null);
    }

    public List<container> listContainers() {
        return new ArrayList<>(containers);
    }
    public Map<sample , sample.sampleID> HashSetSamples(){
        return  Map<>(sample);
    }

    public void updateContainer(long id, String newName, ContainerType newType) {
        container c = getContainer(id);
        if (c == null) throw new IllegalArgumentException("Ошибка: контейнер с id=" + id + " не найден");
        ContainerValidator.validateUpdate(newName, newType);
        if (newName != null) c.setName(newName);
        if (newType != null) c.setType(newType);
    }

    public void removeContainer(long id) {
        container c = getContainer(id);
        if (c == null) throw new IllegalArgumentException("Ошибка: контейнер не найден");
        containers.remove(c);
        // Удаляем связанные слоты и размещения для целостности
        slots.removeIf(s -> s.getContainerId() == id);
        placements.removeIf(p -> p.getContainerId() == id);
    }


    public void createSlots(long containerId, int rows, int cols) {
        if (getContainerById(containerId) == null) {
            throw new IllegalArgumentException("Ошибка: контейнер не найден");
        }
        if (rows <= 0 || cols <= 0) {
            throw new IllegalArgumentException("Ошибка: количество рядов/колонок должно быть > 0");
        }
        if (rows > 1000) {
            throw new IllegalArgumentException("Ошибка: максимальное количество рядов - 1000");
        }
        if (cols > 1000) {
            throw new IllegalArgumentException("Ошибка: максимальное количество колонок - 1000");
        }
        int created = 0;
        for (int r = 1; r <= rows; r++) {
            String rowCode = Slotgenerator.rowToCode(r);

            for (int c = 1; c <= cols; c++) {
                String slotCode = rowCode + c;

                // Проверка дублей внутри контейнера
                boolean exists = slots.stream()
                        .anyMatch(s -> s.getContainerId() == containerId && s.getCode().equals(slotCode));

                if (exists) {
                    throw new IllegalArgumentException("Ошибка: слот " + slotCode + " уже существует");
                }

                slots.add(new slot(idGenerator.getAndIncrement(), containerId, slotCode));
                created++;
            }
        }
    }

    public List<slot> listSlots(long containerId, boolean freeOnly) {
        if (getContainer(containerId) == null)
            throw new IllegalArgumentException("Ошибка: контейнер не найден");
        return slots.stream()
                .filter(s -> s.getContainerId() == containerId && (!freeOnly || !s.isOccupied()))
                .sorted(Comparator.comparing(slot::getCode))
                .collect(Collectors.toList());
    }

    public slot findSlot(long containerId, String code) {
        return slots.stream()
                .filter(s -> s.getContainerId() == containerId && s.getCode().equals(code))
                .findFirst().orElse(null);
    }
// sample manager
    public void createSample (long sampleID, String sampleName, String ownerUsername){
        SampleValidator.validateID(sampleID);
        SampleValidator.validateName(sampleName);
        sample s = new sample(idGenerator.getAndIncrement(), sampleID, sampleName, ownerUsername);
    }

    //  Placement Management
    public void placeSample(long sampleId, long containerId, String slotCode, String owner) {

        if (getContainer(containerId) == null)
            throw new IllegalArgumentException("Ошибка: контейнер не найден");
        slot slot = findSlot(containerId, slotCode);
        if (slot == null) throw new IllegalArgumentException("Ошибка: ячейка не существует");
        if (slot.isOccupied()) throw new IllegalArgumentException("Ошибка: ячейка уже занята");


        removePlacementIfExists(sampleId);

        placements.add(new placement(idGenerator.getAndIncrement(), sampleId, containerId, slot.getId(), owner));
        slot.setOccupied(true);
    }

    public void moveSample(long sampleId, long containerId, String slotCode) {
        placement old = placements.stream().filter(p -> p.getsampleID() == sampleId).findFirst().orElse(null);
        if (old == null) throw new IllegalArgumentException("Ошибка: образец не размещён");
        placeSample(sampleId, containerId, slotCode, old.getOwnerUsername());
    }

    public void removePlacement(long sampleId) {
        placement p = placements.stream().filter(pl -> pl.getSampleId() == sampleId).findFirst().orElse(null);
        if (p == null) throw new IllegalArgumentException("Ошибка: размещение не найдено");
        placements.remove(p);
        slot s = slots.stream().filter(sl -> sl.getId() == p.getSlotId()).findFirst().orElse(null);
        if (s != null) s.setOccupied(false);
    }

    private void removePlacementIfExists(long sampleId) {
        sample.removeIf(smp -> smp.getSampleID() == sampleId);
    }

    public placement findPlacement(long sampleId) {
        return placements.stream().filter(p -> p.getSampleId() == sampleId).findFirst().orElse(null);
    }
    public Map<Character, List<slot>> getContainerMap(long containerId) {
        Map<Character, List<slot>> grid = new TreeMap<>();
        slots.stream()
                .filter(s -> s.getContainerId() == containerId)
                .forEach(s -> grid.computeIfAbsent(s.getCode().charAt(0), k -> new ArrayList<>()).add(s));
        return grid;
    }
}

File: src/main/kotlin/System/StorageService.kt
package System

import Domain.*

class StorageService {

    private val containers = mutableSetOf<Container>()
    private val slots = mutableSetOf<Slot>()
    private val placements = mutableSetOf<Placement>()

    private var containerSeq = 1L
    private var slotSeq = 1L
    private var placementSeq = 1L


    fun addContainer(name: String, type: ContainerType): Container {
        val c = Container(containerSeq++, name, type)
        containers.add(c)
        return c
    }

    fun getContainer(id: Long): Container =
        containers.find { it.id == id }
            ?: throw IllegalArgumentException("Контейнер не найден")

    fun getAllContainers(): List<Container> = containers.toList()


    fun createSlots(containerId: Long, rows: Int, cols: Int) {
        require(rows > 0 && cols > 0) { "Ошибка: rows/cols > 0" }

        for (i in 0 until rows) {
            val row = 'A' + i
            for (j in 1..cols) {
                val code = "$row$j"
                val slot = Slot(slotSeq++, containerId, code)
                slots.add(slot)
            }
        }
    }

    fun getSlots(containerId: Long): List<Slot> =
        slots.filter { it.containerId == containerId }

    fun findSlot(containerId: Long, code: String): Slot =
        slots.find { it.containerId == containerId && it.code == code }
            ?: throw IllegalArgumentException("Слот не найден")


    fun place(sampleId: Long, containerId: Long, code: String) {
        val slot = findSlot(containerId, code)
        require(!slot.occupied) { "Ошибка: слот занят" }

        slot.occupied = true

        val placement = Placement(
            placementSeq++, sampleId, containerId, slot.id
        )
        placements.add(placement)
    }

    fun move(sampleId: Long, containerId: Long, code: String) {
        val old = placements.find { it.sampleId == sampleId }
            ?: throw IllegalArgumentException("Ошибка: не размещён")

        val newSlot = findSlot(containerId, code)
        require(!newSlot.occupied) { "Ошибка: слот занят" }

        slots.find { it.id == old.slotId }?.occupied = false
        newSlot.occupied = true

        placements.remove(old)
        placements.add(
            Placement(placementSeq++, sampleId, containerId, newSlot.id)
        )
    }

    fun remove(sampleId: Long) {
        val p = placements.find { it.sampleId == sampleId }
            ?: throw IllegalArgumentException("Ошибка: не размещён")

        slots.find { it.id == p.slotId }?.occupied = false
        placements.remove(p)
    }

    fun find(sampleId: Long): Placement? =
        placements.find { it.sampleId == sampleId }


    fun getMap(containerId: Long): String {
        val containerSlots = getSlots(containerId)
        val grouped = containerSlots.groupBy { it.code[0] }

        val sb = StringBuilder()
        for ((row, rowSlots) in grouped) {
            sb.append("$row: ")
            for (s in rowSlots.sortedBy { it.code }) {
                sb.append(if (s.occupied) "[X]" else "[ ]")
            }
            sb.append("\n")
        }
        return sb.toString()
    }
}